### PR TITLE
[mlir][spirv] Simplify gpu reduction to spirv logic

### PR DESCRIPTION
**Pull request by kuhar**
Check the type only once and then specialize op handlers based on it.

Do not use boolean types in group arithmetic ops that expect integers.

### CODE REVIEW

**Review by Hardcode84 — APPROVED** [review]

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 1]
```diff
@@ -485,20 +485,22 @@ static std::optional<Value> createGroupReduceOp(OpBuilder &builder,
...
-  using MembptrT = FuncT OpHandler::*;
-  MembptrT handlerPtr;
+  ElemType elementType;
   if (isa<FloatType>(type)) {
-    handlerPtr = &OpHandler::floatFunc;
-  } else if (isa<IntegerType>(type)) {
-    handlerPtr = &OpHandler::intFunc;
+    elementType = ElemType::Float;
+  } else if (auto intTy = dyn_cast<IntegerType>(type)) {
+    elementType = (intTy.getIntOrFloatBitWidth() == 1) ? ElemType::Boolean : ElemType::Integer;
```
[the remark this answers is not in the document]
Comment (kuhar) [reply mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp]: Good point, I added tests. In spir-v booleans are not considered integer types, so we can't use them in most arithmetic ops.

File: mlir/test/Conversion/GPUToSPIRV/reductions.mlir [hunk 2]
```diff
@@ -647,8 +647,108 @@ gpu.module @kernels {
     // CHECK: %{{.*}} = spirv.GroupNonUniformUMax "Subgroup" "Reduce" %[[ARG]] : i32
     %r0 = gpu.subgroup_reduce maxsi %arg : (i32) -> (i32)
     %r1 = gpu.subgroup_reduce maxui %arg : (i32) -> (i32)
+    %r3 = gpu.subgroup_reduce maxui %arg : (i32) -> (i32)
```
Comment (raikonenfnu) [inline mlir/test/Conversion/GPUToSPIRV/reductions.mlir]: Not sure what this new line adds to.

**Review by qedawkins — APPROVED** [review]

**Review by raikonenfnu — APPROVED** [review]
Nice cleanup! :)

File: mlir/test/Conversion/GPUToSPIRV/reductions.mlir (hunk 2 above)
Comment (raikonenfnu) [reply mlir/test/Conversion/GPUToSPIRV/reductions.mlir]: this was resolved.

Comment (kuhar) [reply mlir/test/Conversion/GPUToSPIRV/reductions.mlir]: dropped it in a cleanup commit above

**Pull request merged**